java-checks/src/main/java/org/sonar/java/checks/security/SecureCookieCheck.java

```diff
@@ -19,96 +19,199 @@
  */
 package org.sonar.java.checks.security;
 
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Lists;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
 import org.sonar.java.model.LiteralUtils;
 import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
 import org.sonar.plugins.java.api.JavaFileScannerContext;
 import org.sonar.plugins.java.api.semantic.Symbol;
 import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
 import org.sonar.plugins.java.api.tree.ExpressionTree;
 import org.sonar.plugins.java.api.tree.IdentifierTree;
 import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
 import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
 import org.sonar.plugins.java.api.tree.Tree;
 import org.sonar.plugins.java.api.tree.VariableTree;
 
-import java.util.List;
-
 @Rule(key = "S2092")
 public class SecureCookieCheck extends IssuableSubscriptionVisitor {
 
-  private List<Symbol.VariableSymbol> unsecuredCookies = Lists.newArrayList();
+  private static final String MESSAGE = "Add the \"secure\" attribute to this cookie";
+
+  private static final String JAX_RS_COOKIE = "javax.ws.rs.core.Cookie";
+  private static final String JAX_RS_NEW_COOKIE = "javax.ws.rs.core.NewCookie";
+  private static final String SPRING_SAVED_COOKIE = "org.springframework.security.web.savedrequest.SavedCookie";
+  private static final String PLAY_COOKIE = "play.mvc.Http$Cookie";
+  private static final List<String> COOKIES = Arrays.asList(
+    "javax.servlet.http.Cookie",
+    "java.net.HttpCookie",
+    JAX_RS_COOKIE,
+    JAX_RS_NEW_COOKIE,
+    "org.apache.shiro.web.servlet.SimpleCookie",
+    SPRING_SAVED_COOKIE,
+    PLAY_COOKIE,
+    "play.mvc.Http$CookieBuilder");
+
+  private static final List<String> SETTER_NAMES = Arrays.asList("setSecure", "withSecure");
+
+  /**
+   * Some constructors have the 'secure' parameter and do not need a 'setSecure' call afterwards.
+   */
+  private static final String CONSTRUCTOR = "<init>";
+  private static final String JAVA_LANG_STRING = "java.lang.String";
+  private static final String INT = "int";
+  private static final String BOOLEAN = "boolean";
+
+  private static final List<MethodMatcher> CONSTRUCTORS_WITH_SECURE_PARAM_LAST = Arrays.asList(
+    MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf(JAX_RS_NEW_COOKIE)).name(CONSTRUCTOR)
```

Comment:
I see many repetitions of `MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf(...)).name(CONSTRUCTOR).parameters(...)`

Please extract a dedicated static method:

```java
private static MethodMatcher constructorOf(String type, String ... parameters) {
  return MethodMatcher.create()
    .typeDefinition(TypeCriteria.subtypeOf(type))
    .name(CONSTRUCTOR)
    .parameters(parameters);
}
```

It will simplify all these invocations and hopefully make it easier to read.